KAFKA-8345 (KIP-455): Controller and KafkaApi changes (part 3/4)

stanislavkozlovski opened the pull request.

stanislavkozlovski committed:
[ZK Data] Change TopicZNode structure - encode() method

[Controller] Change ControllerContext#partitionAssignments to store PartitionReplicaAssignment

[ZK Data] Change TopicZNode structure - decode() method
- KafkaZkClient#getPartitionAssignmentForTopics
- KafkaZKClient#getFullReplicaAssignmentForTopics
- Update LeaderAndIsr requests sent by the controller to contain the znode info
- Update the controller to initialize the full data from the znode on failover/startup
- Add a ControllerContextTest
- Adds ControllerContext#partitionFullReplicaAssignment and ControllerContext#updatePartitionFullReplicaAssignment

Change ZkClient#setTopicAssignment to take PartitionReplicaAssignment -- this ensures we never overwrite data about a partition when updating the partition count through AdminManager/tools

[Controller] Handle both ZK and API triggered reassignments + reassignment overrides

Adds a persistedInZk flag to ReassignedPartitionsContext. This lets us keep track of partitions that are stored in the ZK /admin/reassign_partitions znode, even in the case when a reassignment was initially triggered by that znode but overridden by an API call

Rename PartitionReassignment to LegacyZkPartitionReassignment to tell it apart from the future API reassignment event

We handle overrides by reverting any on-going reassignment when a new one for the same partition comes.
To do this, we store two new fields in ReassignedPartitionsContext - originalAssignment (this is removed in a future commit) and ongoingReassignment
  - We use ongoingReassignment to know when (and what) we need to revert
  - We use original assignment to know what to revert to. It is important we use the original assignment to know the exact order the replicas were in. We will further use this in the next commits to cancel ongoing reassignments

 [RPC] Implement ListPartitionReassignments logic

Adds a ListPartitionReassignment controller state
For the List() and List(partitions) API, we return only those partitions which are going through a reassignment.

[RPC] Implement AlterPartitionReassignments logic

Add 5 new ReassignPartitionsClusterTest tests for:
- API Alter Reassignment
- API List Reassignment during a reassignment
- znode reassignment overriding an API reassignment
- reassignment znode deleted before whole API reassignment finishes
- API reassignment stacking
Add one AdminClientIntegrationTest for the UNNOWN_TOPIC_OR_PARTITION error in the Alter API

Also improve reassignment revert logic
- Fix cleanup reassignment step (when reverting) to properly populate the replica and adding replica sets. It now also updates ZK
- Do not store previous (original) assignment in cache to revert to when a revert is due. This loses ordering but is consistent with the cases where the controller fails

[AlterReassignments] Validate against invalid replicas

stanislavkozlovski commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+   *   . URS denotes the unnecessary replicas, ones which are currently part of the ongoing reassignment but will not be part of the new one
+   *   . OVRS denotes the overlapping replica set - replicas which are part of the ongoing reassignment and will be part of the overriding reassignment
+   *       (it is essentially (RS - ORS) - URS)
+   *   1 Set RS = ORS + OVRS, AR = OVRS, RS = [] in memory
```
Just a quick comment to make things super clear:

Reassignment #1
```
ORS = [1,2,3]
TRS = [3,4,5]
```
Reassignment #2 (overriding)
```
TRS = [3,5,6]
OVRS = [5]
URS = [4]
```

stanislavkozlovski commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -605,36 +691,42 @@ class KafkaController(val config: KafkaConfig,
         } else {
             error(s"Ignoring request to reassign partition $tp that doesn't exist.")
             partitionsToBeRemovedFromReassignment.add(tp)
+            reassignmentResults.put(tp, new ApiError(Errors.UNKNOWN_TOPIC_OR_PARTITION, "The partition does not exist."))
```
Do we want to guarantee that if controllerContext.partitionReplicaAssignment(tp).isBeingReassigned then controllerContext.partitionsBeingReassigned(tp).isPresent?
At this point, I don't know if we should update the in-memory and ZK replica assignment to one that is not being reassigned. Since we remove the partition from the partitionsBeingReassigned collection, it seems to make sense to also update in-memory and ZK to truly treat the reassignment as failed.
Yet, do we need to send LeaderAndIsr requests if we change those?

The one scenario where this divergence could happen (controllerContext.partitionReplicaAssignment(tp).isBeingReassigned is true and controllerContext.partitionsBeingReassigned(tp).isPresent is not)
is if we have an UNKNOWN_SERVER_ERROR during the reassignment (line 723)

cmccabe replied:
I think we should try to avoid having an in-memory state that does not match the ZK state.  Can you think of a case where that might happen here?

stanislavkozlovski replied:
I think that might happen if we throw an error somewhere inside `onPartitionReassignment` after it updates the in-memory state. This wasn't handled previously either

The question here is more about the two separate in-memory collections diverging - `partitionsBeingReassigned` and `partitionReplicaAssignment`. If we get an unknown error inside `onPartitionReassignment` we may have updated the in-memory `partitionReplicaAssignment` but in `maybeTriggerPartitionReassignment` we catch the error and remove the partition from `partitionsBeingReassigned`. So we may end up with an in-memory state indicating a reassignment but none scheduled due to some failure.

I can't think of a great way to guard against this. It may be good to leave it as is

cmccabe replied:
It's difficult to handle arbitrary exceptions propagating from any line of code.  If we're confident that we've carefully checked the code in onPartitionReassignment, let's not block on this for now.

stanislavkozlovski commented on core/src/main/scala/kafka/controller/ControllerContext.scala:
```diff
@@ -80,14 +110,55 @@ class ControllerContext {
     replicaStates.clear()
   }
 
+  def partitionReplicaAssignment(topicPartition: TopicPartition): Seq[Int] = {
```
I intentionally kept the name of this method the same to make it easier to review (keep the diff smaller), ditto for the update methods below. Perhaps we could rename them to `partitionReplicas/updatePartitionReplicas` and the new methods to `partitionAssignment/updatePartitionAssignment`

stanislavkozlovski commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+   *       (it is essentially (RS - ORS) - URS)
+   *   1 Set RS = ORS + OVRS, AR = OVRS, RS = [] in memory
+   *   2 Send LeaderAndIsr request with RS = ORS + OVRS, AR = [], RS = [] to all brokers in ORS + OVRS
+   *     (because the ongoing reassignment is in phase A, we know we wouldn't have a leader in URS)
```
Technically we could have a leader in URS if preferred leader election was triggered while the reassignment was happening

cmccabe replied:
Can you please add this to the JavaDoc comment?

stanislavkozlovski commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+   *     (because the ongoing reassignment is in phase A, we know we wouldn't have a leader in URS)
+   *   3 Replicas in URS -> Offline (force those replicas out of ISR)
+   *   4 Replicas in URS -> NonExistentReplica (force those replicas to be deleted)
+   *   5 Update ZK with RS = ORS + OVRS, AR = OVRS, RS = []
```
I am not entirely sure we need this last update but I leaned on the side of caution

stanislavkozlovski commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+    maybeRevertOngoingReassignment(topicPartition, reassignedPartitionContext)
+
+    val partitionAssignment = PartitionReplicaAssignment.fromOldAndNewReplicas(
+      oldReplicas = controllerContext.partitionFullReplicaAssignment(topicPartition).previousAssignment.replicas,
```
We need to use the original replica set here. In case of an on-going reassignment, the last step in `maybeRevertOngoingReassignment` is
```
// Update ZK with RS = ORS + OVRS, AR = OVRS, RR = [].
```
To get the ORS, we need to call `#previousAssignment.replicas`

Since `controllerContext.partitionFullReplicaAssignment(topicPartition).previousAssignment == controllerContext.partitionFullReplicaAssignment(topicPartition)` when there is no on-going reassignment, we call it all the time to more easily handle both cases.

stanislavkozlovski commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+    val partitionAssignment = PartitionReplicaAssignment.fromOldAndNewReplicas(
+      oldReplicas = controllerContext.partitionFullReplicaAssignment(topicPartition).previousAssignment.replicas,
+      newReplicas = reassignedPartitionContext.newReplicas)
+    assert(reassignedPartitionContext.newReplicas == partitionAssignment.targetReplicas,
```
I've added in `assert()` calls throughout the code -- they helped me catch some bugs. I am wondering whether it's worth it to keep them for one AK release to catch issues we didn't expect or to remove them. I know we've had some before but I think I've heard we're moving away from them

cmccabe replied:
`assert` is turned off by certain JVM flags.  So it's not really helpful for catching errors, since it is often turned off.

Considering that we already have this problem in the controller, it may be worth doing the conversion to something else en masse, rather than worrying about it in this PR.

stanislavkozlovski commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -878,15 +996,41 @@ class KafkaController(val config: KafkaConfig,
+
+  private def removePartitionsFromZkReassignment(partitionsToBeRemoved: Set[TopicPartition]): Unit = {
+    if (!zkClient.reassignPartitionsInProgress()) {
+      debug(s"Cannot remove partitions $partitionsToBeRemoved from ZooKeeper because there is no ZooKeeper reassignment present")
```
Come to think of it this would get called on API every reassignment so it may be worth to not log it?

cmccabe replied:
I think it's OK to keep the log.

stanislavkozlovski commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2298,20 +2299,85 @@ class KafkaApis(val requestChannel: RequestChannel,
+          val tp = new TopicPartition(reassignableTopic.name(), reassignablePartition.partitionIndex())
+          if (reassignablePartition.replicas() == null)
+            tp -> None // revert call
+          else
```
I've opted out of having the INVALID_REASSIGNMENT error. Mainly because it's tricky to know when a replica is intermittently offline, as the znode and controller memory cache store only the online brokers.

stanislavkozlovski replied:
We could have some preliminary checks here for an empty list or negative IDs though

stanislavkozlovski replied:
We discussed here and I changed my mind that it's a goo idea to have - https://github.com/apache/kafka/pull/7120#discussion_r308861346

stanislavkozlovski commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -238,19 +239,55 @@ object TopicsZNode {
-    Json.encodeAsBytes(Map("version" -> 1, "partitions" -> assignmentJson.asJava).asJava)
+
+    Json.encodeAsBytes(Map(
+      "version" -> 2,
```
Do we need to bump the version here?

stanislavkozlovski commented on core/src/main/scala/kafka/controller/ControllerContext.scala:
```diff
@@ -80,14 +110,55 @@ class ControllerContext {
+    partitionAssignments.getOrElse(topicPartition.topic, mutable.Map.empty)
+      .get(topicPartition.partition) match {
+      case Some(partitionAssignment) => partitionAssignment
+      case None => PartitionReplicaAssignment(Seq(), Seq(), Seq())
```
Is there any benefit to this being PartitionReplicaAssignment.EMPTY? It's probably easier to read

cmccabe replied:
It's fine either way.

stanislavkozlovski commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsClusterTest.scala:
```diff
@@ -12,33 +12,37 @@
+import org.apache.kafka.common.config.ConfigResource
+import org.apache.kafka.common.errors.NoReassignmentInProgressException
 
 class ReassignPartitionsClusterTest extends ZooKeeperTestHarness with Logging {
```
I ran the entire test suite locally 37 times in a row with no failures whatsoever. The new tests do not seem to be flaky

stanislavkozlovski commented:
Will rebase with trunk once https://github.com/apache/kafka/pull/7120 is merged.
Excited to see a fully-green build! :)

stanislavkozlovski commented:
A short summary:

We use the same ReassignPartitions event to handle both the ZK and API reassignment. I think this approach reuses more code and is more easily read when trying to understand the interactions between both
We use controller events to handle the API calls.
Colin's original draft implementation had a CallbackManager for managing callbacks from and into the controller.
I've decided to not implement it here because:
I don't completely understand how it would work - Colin's implementation didn't have it plugged in to anything yet
It would essentially make the controller multi-threaded if I understand it correctly. I'd like to reduce the complexity and scope of this PR.
I believe that we can delegate it to a separate PR/JIRA, that way we could also move other parts of the controller to leverage it (e.g the newly-added ElectLeaders API)

cmccabe commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2311,26 +2312,85 @@ class KafkaApis(val requestChannel: RequestChannel,
+      }
+
+      sendResponseMaybeThrottle(request, requestThrottleMs =>
+        new AlterPartitionReassignmentsResponse(responseData.setThrottleTimeMs(requestThrottleMs).toStruct(0))
```
The version that we're writing should be matched to the version that we received.  So rather than hard-coding 0, we should look at the message header.

stanislavkozlovski replied:
oops

cmccabe commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2311,26 +2312,85 @@ class KafkaApis(val requestChannel: RequestChannel,
+
+    val partitionsOpt = listPartitionReassignmentsRequest.data().topics() match {
+      case topics: Any =>
+        Some(topics.iterator().asScala.flatMap { topic =>
```
is flatMap really needed here, or should it be map?

stanislavkozlovski replied:
It's not needed

stanislavkozlovski replied:
Should be `flatMap` since we do another map inside on the partitions indices to convert to TopicPartition and we want a big list of TopicPartitions, rather than one for each topic

cmccabe replied:
thanks for the explanation

cmccabe commented:
Hi @stanislavkozlovski. Thanks for working on this.  It looks like it needs a rebase on trunk.

In the Scala code, you made good use of `groupBy` and `distinct`... very nice.  I still forget about those sometimes.

I remember earlier we talked about having a separate thread to handle formatting the responses being sent back via `KafkaApis`.  I can see this patch leaves that out, in favor of just having the controller thread do that work.  I guess that is simpler.  We will need to think about how to handle timeouts, though.  It seems like if the controller thread is busy, we won't get around to sending back a response indicating a timeout.  So perhaps we will have to give up on idea of a request-specified timeout in `AlterPartitionReassignmentsRequest` for now, or add a background thread here.

Now that we have `ControllerState#ListPartitionReassignment`, does it make sense to rename `ControllerState#PartitionReassignment` to `ControllerState#AlterPartitionReassignment` for consistency's sake?

stanislavkozlovski commented:
Thanks for the review. Good point that we currently completely ignore the timeout, I will think a bit more on this and get back to you

stanislavkozlovski commented:
I had to rebase with master. It was a big rebase and I didn't have time to fully go over everything again, so there may be some leftover. It did compile and pass checkstyle locally though so I assume it's good

stanislavkozlovski commented:
Note that I need to change the `preferredReplicaImbalanceCount` metric to report a preferred leader imbalance if and only if the preferred replica is in ISR. This would avoid reporting the metric at the very start of a rebalance where a new replica which isn't in ISR yet is the first preferred one.

stanislavkozlovski committed:
Address PR comments and add a note on upgrading

stanislavkozlovski committed:
Modify preferredReplicaImbalanceCount to not count reassigning replicas until they join ISR

stanislavkozlovski commented:
I have removed the `TimeoutMs` fields from ListPartitionReassignmentsRequest and  AlterPartitionReassignmentsRequest. It's right that if the controller thread is busy, we won't get around to sending back a response. While this means that we don't really have a way to avoid having open sockets pile up if the controller thread hangs, we most likely have bigger problems if that thread hangs.

I renamed ControllerState#PartitionReassignment to ControllerState#AlterPartitionReassignment, thanks for catching that.

With regards to the version in the topics znode, we've decided to bump it to 2 but not have it depend on the inter.broker.protocol.

@cmccabe I also added a new change (commit 7417ec31ddcd008c7cd42280b4d55309d37b361f) where I changed the semantics of the PreferredReplicaImbalanceCount metric a bit. I changed it to not count partitions whose newly-added, still reassigning preferred replica is not the leader. I had discussed this a bit with @hachikuji offline

cmccabe commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+   * RS = current assigned replica set
+   * ORS = Original replica set for partition
+   * TRS = Reassigned (target) replica set
+   * AR = The replica we are adding as part of this reassignment
```
Should be "replica(s)", because it may be more than one (or may be zero)

cmccabe commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+   * ORS = Original replica set for partition
+   * TRS = Reassigned (target) replica set
+   * AR = The replica we are adding as part of this reassignment
+   * RR = The replica we are removing as part of this reassignment
```
Should be "replica(s)", because it may be more than one

cmccabe commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+   * AR = The replica we are adding as part of this reassignment
+   * RR = The replica we are removing as part of this reassignment
+   *
+   * We can explain a reassignment in two consecutive phases and a possibly-prerequisite cleanup phase, each with its own steps:
```
This is confusing.  Just write "a reassignment may have up to three phases"

cmccabe commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
-   * {1,2,3,4,5,6}      4/{1,2,3,4,5,6}     (step 7)
-   * {1,2,3,4,5,6}      4/{4,5,6}           (step 8)
-   * {4,5,6}            4/{4,5,6}           (step 10)
+   * Phase A: Initial trigger (when RAR != ISR)
```
What is "RAR"?  Is this supposed to be TRS?

stanislavkozlovski replied:
Yes, good catch

cmccabe replied:
There are still a few cases where we're referring to RAR rather than TRS

cmccabe commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+   *     (because the ongoing reassignment is in phase A, we know we wouldn't have a leader in URS)
+   *   3 Replicas in URS -> Offline (force those replicas out of ISR)
+   *   4 Replicas in URS -> NonExistentReplica (force those replicas to be deleted)
+   *   5 Update ZK with RS = ORS + OVRS, AR = OVRS, RS = []
```
This change should involve a bump of of the leader epoch, the same as the one in phase A, right?

stanislavkozlovski replied:
The leader epoch bump is done in step 2 (KafkaController#645)

cmccabe commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+   *   B2. Set RS = TRS, AR = [], RR = [] in memory.
+   *   B3. If the leader is not in TRS, elect a new leader from TRS. If new leader needs to be elected from TRS, a LeaderAndIsr
+   *       will be sent. If not, then leader epoch will be incremented in zookeeper and a LeaderAndIsr request will be sent.
+   *       In any case, the LeaderAndIsr request will have RS = TRS. This will prevent the leader from adding any replica in
```
The leader epoch needs to be incrementing in both cases, right?  If a new leader needs to be elected, and also if a new leader doesn't need to be elected.  So we should just add this as a step.

cmccabe commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2311,26 +2312,85 @@ class KafkaApis(val requestChannel: RequestChannel,
+      }
+
+      sendResponseMaybeThrottle(request, requestThrottleMs =>
+        new AlterPartitionReassignmentsResponse(responseData.setThrottleTimeMs(requestThrottleMs).toStruct(request.header.apiVersion))
```
There is no reason to convert the responseData to a Struct before creating the AlterPartitionReassignmentsResponse.  There is a constructor overload for AlterPartitionReassignmentsResponse that can take the data object directly.  Converting to and from Struct is inefficient, and it increases our dependency on Struct, which we want to phase out eventually (in favor of direct serialization).

stanislavkozlovski replied:
Thanks. I'm not sure where I got this from at all, since nothing else in KafkaApis uses `toStruct()`

cmccabe commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -508,7 +508,8 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
    * @throws KeeperException if there is an error while creating assignment
```
On line 505, we should explain how this function is different from the one that returns PartitionReplicaAssignment.  Something like "this function does not return replicas which are being added or removed in a reassignment"

stanislavkozlovski replied:
Nothing in this class returns PartitionReplicaAssignment, are you referring to the parameters we accept?

stanislavkozlovski replied:
The function on line 505 (`createTopicAssignment`) doesn't return anything. It creates an assignment. I don't think it's worth distinguishing between reassignments here because we're creating a new topic

cmccabe replied:
yes, I agree, the JavaDoc for the `createTopicAssignment` functions is probably fine as-is.

cmccabe commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -569,18 +570,27 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
 
   /**
-   * Gets the assignments for the given topics.
+   * Gets the replica assignments for the given topics.
```
Same thing here.  We should explain that this doesn't handle reassigning replicas

cmccabe commented:
On ZkData.scala, on line 408, we should add an explanation of what ReassignPartitionsZNode is, and the fact that it was deprecated by KIP-455.
(The terrible github review UI doesn't let me leave a comment on this line, since you didn't change it)

cmccabe commented:
KafkaZkClient: for the deprecated functions, please add a short explanation of why they're deprecated, and what the new API that should be used is.

cmccabe commented:
I think let's keep `timeoutMs` in the message format for now.  We are going to implement a general way of doing timeouts for controller RPCs in the future as part of the KIP-500 work.  For now, let's just file a follow-on JIRA to implement this timeout.

stanislavkozlovski committed:
Address PR comments about improving javadocs

stanislavkozlovski committed:
Bring back timeoutMs to the protocol

stanislavkozlovski commented:
I think I addressed most of the comments. I opened https://issues.apache.org/jira/browse/KAFKA-8873 for the timeout implementation and mentioned it in the KIP

stanislavkozlovski committed:
Remove left-over usage of RAR in comments

stanislavkozlovski committed:
Update comments to match the code

stanislavkozlovski committed:
Update memory with RS = ORS + TRS, AR = TRS - ORS and RR = ORS - TRS only in phase A, right after ZK update

The reason we updated it before both cases was just because the phase B function `moveReassignedPartitionLeaderIfRequired` read from what's stored in memory. Updated the function to take in a parameter of the assignment we want it to work with.

This handles cases where the TRS is in ISR (we skip phase A) but we want to send the LeaderAndIsr to more than TRS. e.g R=[1,2,3,4], ISR=[1,2,3,4], TRS=[1,2,3]

stanislavkozlovski committed:
Send LAIR to target replicas where possible

stanislavkozlovski committed:
Remove unnecessary ZK write in revert step

stanislavkozlovski committed:
Refactor the code a bit to stop relying on the update of `controllerContext` structure update

stanislavkozlovski commented:
I think this is ready for review again.

I double-check we bump and save the leader epoch in ZK before sending a LeaderAndIsr request

We discussed offline how altering ZK after memory is fair in certain cases - specifically in phase B and the cleanup phase, due to needing to make sure we at least send StopReplica to the retired replicas

I modified the code to send LeaderAndIsr only to the replicas present in the LAIR when possible. I am still a bit confused by the partition state machine code - but it may send the request to more replicas due to the fact that it reads the replica set from ZK to determine where to send a request. I don't really feel comfortable changing that code yet, so I'd rather have us send to more than necessary, like we always have, for now.

stanislavkozlovski commented:
The two JDK 11/Scala 2.13 failures are unrelated:
```
org.apache.kafka.streams.KafkaStreamsTest.statefulTopologyShouldCreateStateDirectory
org.apache.kafka.streams.KafkaStreamsTest.inMemoryStatefulTopologyShouldNotCreateStateDirectory
```

viktorsomogyi commented on core/src/main/scala/kafka/controller/ControllerContext.scala:
```diff
@@ -22,6 +22,45 @@ import org.apache.kafka.common.TopicPartition
+    */
+  def previousAssignment: PartitionReplicaAssignment = {
+    PartitionReplicaAssignment(
+      replicas.filterNot(addingReplicas.contains(_)),
```
Why don't we add removingReplicas too in order to fully restore the previous state?

stanislavkozlovski replied:
Any previous assignment would not be reassigning - it would not have adding/removing replicas.
We always revert the on-going reassignment when a new one is triggered.

viktorsomogyi commented on core/src/main/scala/kafka/controller/ControllerContext.scala:
```diff
@@ -22,6 +22,45 @@ import org.apache.kafka.common.TopicPartition
+    * Returns the target replica assignment for this partition.
+    * This is different than the `replicas` variable only when there is a reassignment going on
+    */
+  def targetReplicas: Seq[Int] = replicas.filterNot(removingReplicas.contains(_))
```
Similarly to this one: would you please summarize why don't we take original + adding_replicas - removing_replicas as target? Does this have something to do with the2 phase algorithm that first we add AR and then remove RR? (And apologies if I missed something obvious.)

stanislavkozlovski replied:
You're misunderstanding the `replicas` set. It is not the original replicas. Let me give a quick example:

replicas = [1,2,3]
request with target_replicas = [4,5,6]
Upon processing the request, our PartitionReplicaAssignment class would look like this:
```
PartitionReplicaAssignment(
  replicas=[1,2,3,4,5,6]
  addingReplicas=[4,5,6]
  removingReplicas=[1,2,3]
)
```
You can see where we create the case class instance in `onPartitionReassignment`, specifically the `PartitionReplicaAssignment.fromOldAndNewReplicas` method which helps us create the new assignment.

viktorsomogyi replied:
Thanks, this is helpful.

viktorsomogyi commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1425,25 +1589,134 @@ class KafkaController(val config: KafkaConfig,
+      case Some(targetReplicas) =>
+        if (reassignmentIsInProgress) {
+          info(s"Overriding old reassignment for partition $partition " +
+            s"(with target replicas ${replicaAssignment.targetReplicas.mkString(",")}) " +
```
nit: can `replicaAssignment.targetReplicas` be different than `targetReplicas`? If not this log statement might be a bit redundant.

stanislavkozlovski replied:
Yes, that's the reason of the log. It is called when we are overriding an existing reassignment. i.e an on-going reassignment is happening (`replicaAssignment`) and we call the function to save a new reassignment (`targetReplicas`)

They can be the same, sure, if the user sends the same request twice in a row - but it is worth logging regardless. At the very least, it would help diagnose bugs easier. Besides that, I think it's a very useful log

viktorsomogyi reviewed (commented):
Hey Stan, started to review your changes, I'll probably go through this during the week and incrementally add comments where needed to.

stanislavkozlovski committed:
Remove stale comment, re-populate memory with ZK assignment state in unexpected exceptions and log when leader elections are retried

cmccabe commented:
@stanislavkozlovski: I think this is almost ready!

We should probably add logging for when a partition leader election fails and we retry.  This will let us know if we're in an infinite partition leader election retry loop (this is an existing logging gap...)

I'm curious whether the call to `onPartitionReassignment` in `onBrokerStartup` is still needed.

@viktorsomogyi : thanks for reviewing this.  We were hoping to get it in soon so we could start testing.  Having this in will also unblock the follow-on work like KIP-352.  I promised Stan I would get it in last week, but I ran out of time.

So I guess let's give it one or two days and then commit, if we don't find any blockers.  And we can continue to incrementally improve this in follow-on changes.

cmccabe commented:
LGTM.  Thanks, @stanislavkozlovski
